Text/CustomEntitySearch/CustomEntitySearch.cs

```diff
@@ -52,48 +53,273 @@ public static class CustomEntitySearch
+                                 AddValues(word, text, entities, entitiesFound, leniency, caseSensitive);
+                                 if (synonyms != null && synonyms.ContainsKey(word))
+                                 {
+                                     foreach (string synonym in synonyms[word])
+                                     {
+                                         leniency = (exactMatches != null && exactMatches.Contains(synonym)) ? 0 : offset;
+                                         AddValues(synonym, text, entities, entitiesFound, leniency, caseSensitive);
+                                     }
+                                 }
+                             }
+                         }
+
+                         outRecord.Data["Entities"] = entities;
+                         outRecord.Data["EntitiesFound"] = entitiesFound;
+                     }
+                    return outRecord;
+                });
+
+            return new OkObjectResult(response);
+        }
+
+        public static void AddValues(string checkMatch, string text, List<Entity> entities, HashSet<string> entitiesFound, int leniency, bool caseSensitive)
+        {
+            bool addWord = false;
+            string escapedWord = Regex.Escape(checkMatch);
+            string pattern = (caseSensitive) ? @"\b(?x:" + escapedWord + @")\b" : @"\b(?ix:" + escapedWord + @")\b";
+            MatchCollection entityMatch = Regex.Matches(text, pattern, RegexOptions.Compiled, TimeSpan.FromSeconds(MaxRegexEvalTime));
+            if (entityMatch.Count != 0)
+            {
+                foreach (Match match in entityMatch)
+                {
+                    entities.Add(
+                        new Entity
+                        {
+                            Category = "customEntity",
+                            Value = match.Value,
+                            Offset = match.Index,
+                            Confidence = 1
+                        });
+                }
+                addWord = true;
+            }
+
+            if (leniency > 0)
+            {
+                // using KMP for now
+                // Create Table
+                IList<int> KMPTable = (caseSensitive) ? CreateKMPTable(checkMatch) : CreateKMPTable(checkMatch.ToLower());
+
+                // Begin searching!
+                int currTextChar = 0;
+                int currWordChar = 0;
+                StringWriter wordFound = new StringWriter();
+                double currMismatch = 0;
+                IList<char> wordCharArray = (caseSensitive) ? checkMatch.ToCharArray() : checkMatch.ToLower().ToCharArray();
+                IList<char> textCharArray = (caseSensitive) ? text.ToCharArray() : text.ToLower().ToCharArray();
+                IList<string> valuesList = new List<string>();
+                IList<int> offsetList = new List<int>();
+                IList<double> confidenceList = new List<double>();
```

Comment:
Could you just hve one list of entites here rather than 3 different lists that you have to stitch back together?

Reply:
Yes!